src/config.cpp

```diff
@@ -0,0 +1,322 @@
+                valueStr.replace(startPos, varName.length(), varValue);
+                value = valueStr;
+            }
+        }
+    }
+
+    return jsonFlat.unflatten();
+}
+
+
+std::map<std::string, std::string> fillComponents(const nlohmann::json& json,
+                                                  const PathResolver& resolver) {
+    const auto components = json.at("components");
+    std::map<std::string, std::string> result;
+
+    for (auto it = components.begin(); it != components.end(); ++it) {
+        result[it.key()] = resolver.toAbsolute(it.value());
+    }
+
+    return result;
+}
+
+struct SubnetworkFiles {
+    std::string elements;
+    std::string types;
+};
+
+std::vector<SubnetworkFiles> fillSubnetwork(nlohmann::json& networks,
+                                            const std::string& prefix,
+                                            const PathResolver& resolver) {
+    std::vector<SubnetworkFiles> output;
+
+    const std::string component = prefix + "s";
+    const std::string elementsFile = prefix + "s_file";
+    const std::string typesFile = prefix + "_types_file";
+
+    const auto iter = networks.find(component);
+    if (iter == networks.end()) {
+        return output;
+    }
+
+    for (const auto& node : *iter) {
+        auto h5File = resolver.toAbsolute(node.at(elementsFile));
+        auto csvFile = node.at(typesFile).is_null() ? std::string()
+                                                    : resolver.toAbsolute(node.at(typesFile));
+        output.emplace_back(SubnetworkFiles{h5File, csvFile});
+    }
+
+    return output;
+}
+
+using Variables = std::map<std::string, std::string>;
+
+Variables readVariables(const nlohmann::json& json) {
+    Variables variables;
+
+    if (json.find("networks") == json.end()) {
+        return variables;
+    }
+
+    const auto manifest = json["manifest"];
+
+    const std::regex regexVariable(R"(\$[a-zA-Z0-9_]*)");
+
+    for (auto it = manifest.begin(); it != manifest.end(); ++it) {
+        const auto name = it.key();
+
+        if (std::regex_match(name, regexVariable)) {
+            variables[name] = it.value();
+        } else {
+            throw SonataError(fmt::format("Invalid variable `{}`", name));
+        }
+    }
+
+    return variables;
+}
+
+nlohmann::json parseSonataJson(const std::string& contents) {
+    const auto json = nlohmann::json::parse(contents);
+
+    const auto vars = replaceVariables(readVariables(json));
+    return expandVariables(json, vars);
+}
+
+
+template <typename Population>
+std::map<std::string, SubnetworkFiles> resolvePopulations(
+    const std::vector<SubnetworkFiles>& networkNodes) {
+    using PopulationStorage = bbp::sonata::PopulationStorage<Population>;
+    std::map<std::string, SubnetworkFiles> result;
+
+    for (const auto& network : networkNodes) {
+        const auto population = PopulationStorage(network.elements, network.types);
+        for (auto name : population.populationNames()) {
```

Comment:
Any reason not to:
```suggestion
        const auto& population = PopulationStorage(network.elements, network.types);
        for (const auto& name : population.populationNames()) {
```

Reply:
No reason at all, I replaced the values by `const &` on the for range loops. In the population statement, though, I don't understand the reference of a assigned constructed object. However, I have replace the assignment with with in-place construction to make the code cleaner.